core/src/main/java/org/apache/calcite/rel/rules/SortJoinTransposeRule.java

```diff
@@ -174,6 +177,24 @@ public SortJoinTransposeRule(Class<? extends Sort> sortClass,
     call.transformTo(sortCopy);
   }
 
+  /**
+   * Returns the fetch value for the inner sort when pushing sort past join.
+   * The value is outer sort's offset + fetch.
+   *
+   * @param sort the outer sort
+   * @param rexBuilder RexBuilder to create literals
+   * @return fetch for inner sort
+   */
+  private static @Nullable RexNode calculateInnerSortFetch(Sort sort, RexBuilder rexBuilder) {
+    if (sort.fetch == null) {
+      return null;
+    }
+    final int outerFetch = RexLiteral.intValue(sort.fetch);
```

Comment:
why not use BigInteger for these calculations?

Reply:
This doesn't mean the previous way of doing things is right...
You can write a test with a very large limit to see what happens. In the best case you get an exception. In the worst case you get silent wraparound.

Reply:
I have created another JIRA [CALCITE-7154](https://issues.apache.org/jira/browse/CALCITE-7154) to document this issue.